### PR TITLE
Added code to replicate the Sherwen Halogen mechanism in F0AM

### PR DESCRIPTION
**Pull request by paulsromer**
Included in this commit is the  Sherwen mechanism in F0AM format and my modifications to F0AM to make it work. The Sherwen mechanism itself is in two files, GEOSChemHalogenReactions.m and GEOSChemHalogenUptake_Moderate.m. Other modifications include calculations of the photolysis rates using the hybrid approach and an example file of the halogen setup. There were also three heterogeneous reactions that depend on the composition of the particle, which was more in-depth than I was willing to do. Instead, I followed Parrella et al., 2012, which treated both species as gas phase, and calculated the reactive uptake rate for the limitting reactant. But making that work  required some ugly hacks to initialize_chem, dydt_eval, and JAC_eval, so there's probably a better way to make that work in the long-term.

### CODE REVIEW

File: Chem/Photolysis/Extend_HybridJ.m [hunk 1]
```diff
@@ -0,0 +1,130 @@
+%ExtendHybridJ.m
```
Comment (gmwolfe) [inline Chem/Photolysis/Extend_HybridJ.m]: In the next release, will can add all of these to the standard J-value list. I will also include a tutorial and all code/solar spectra needed for users to generate their own "hybrid" J-values.

File: Core/InitializeChemistry.m [hunk 2]
```diff
@@ -131,7 +131,13 @@
 
 Gstr = Gstr(1:i,:);
 iempty = cellfun('isempty',Gstr);
-Gstr(iempty) = {'ONE'}; %replace empty cells with name of "ONE" species (all 1's)
+Gstr(iempty) = {'ONE'}; %replace empty cells with name of "ONE" species (all 1's)\
+%%%%% Add in code to deal with limiting reactants
+% PSR 190109
+[r, c] = find(~cellfun(@isempty,strfind(Gstr,'limiting')));
+Gstr = strrep(Gstr,'limiting_','');
```
Comment (gmwolfe) [inline Core/InitializeChemistry.m]: Probably cleaner to deal with these special "limiting reagent" reactions by adding a boolean flag that can be thrown for individual reactions within chemistry scripts. Then the Gstr values in reactions would have normal names and this bit of code could be deleted.

File: Core/InitializeChemistry.m [hunk 3]
```diff
@@ -131,7 +131,13 @@
 
 Gstr = Gstr(1:i,:);
 iempty = cellfun('isempty',Gstr);
-Gstr(iempty) = {'ONE'}; %replace empty cells with name of "ONE" species (all 1's)
+Gstr(iempty) = {'ONE'}; %replace empty cells with name of "ONE" species (all 1's)\
+%%%%% Add in code to deal with limiting reactants
```
Comment (gmwolfe) [inline Core/InitializeChemistry.m]: Probably cleaner to deal with these special "limiting reagent" reactions by adding a boolean flag that can be thrown for individual reactions within chemistry scripts. Then the Gstr values in reactions would have normal names and this bit of code could be deleted.

File: Core/F0AM_ModelCore.m [hunk 4]
```diff
@@ -339,6 +446,11 @@
 %Chemical rates
 Conc(:,2) = sum(Conc(:,Chem.iRO2),2);
 G = Conc(:,Chem.iG(:,1)).*Conc(:,Chem.iG(:,2));
+%Need to adjust G here as well to deal with limiting reactants in heterogeneous
```
Comment (paulsromer) [inline Core/F0AM_ModelCore.m]: This section fixes the output values of Rates to be the correct value for reactions with limiting reagents.

**Comment by gmwolfe** [comment]
Github is breaking my brain. B/c these mods were made with F0AMv3.2, there are multiple conflicts with the current version in the guttywhats. I don't want to break Paul's code, so I just made a new branch off of F0AMv4 where I can merge things in manually. This makes more logical sense to me, anyway.

**Comment by gmwolfe** [comment]
I have added this mechanism to F0AMv4 as best I could, but it ain't perfect. The family conservation does not seem to work well with the halogen families, and I do not understand why. Possibly something to do with the stiffness of the family, or else the Jacobian still isn't coded right. In any event, if someone wants to replicate Paul's work I suggest they just use his v3.2 setup for now.

**Pull request closed without merging**